Stabilize slice::strip_prefix and slice::strip_suffix

ijackson opened the pull request:
These two methods are useful.  The corresponding methods on `str` are already stable.

I believe that stablising these now would not get in the way of, in the future, extending these to take a richer pattern API a la `str`'s patterns.

Tracking PR: #73413.  I also have an outstanding PR to improve the docs for these two functions and the corresponding ones on `str`: #75078

I have tried to follow the [instructions in the dev guide](https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#stabilization-pr).  The part to do with `compiler/rustc_feature` did not seem applicable.  I assume that's because these are just library features, so there is no corresponding machinery in rustc.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @Mark-Simulacrum (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://rustc-dev-guide.rust-lang.org/contributing.html) for more information.

Amanieu commented:
@rfcbot fcp merge

This API seems fine for now, but https://github.com/rust-lang/rfcs/pull/2500 plans to generalize them to take a `Pattern` as an argument instead.

Mark-Simulacrum commented:
That seems like it will be at least an inference break - do we want to try to future proof these somehow?

KodrAus commented:
The stabilized `str` variants [expose the unstable `Pattern` trait](https://doc.rust-lang.org/std/primitive.str.html#method.strip_prefix). Is there anything stopping us from doing that here too?

lzutao commented:
@KodrAus  `std::str::pattern::Pattern` is str oriented, .i.e. it only supports string type.
There are plans to generalize Pattern to other types. But those are [not implemented](https://github.com/rust-lang/rust/pull/76901/).

ijackson commented:
Mark Rousskov writes ("Re: [rust-lang/rust] Stabilize slice::strip_prefix and slice::strip_suffix (#77853)"):
> That seems like it will be at least an inference break

Good point.  Thanks for pointing this out.

> - do we want to try to future proof these somehow?

Yes, surely.

What if we provided a slice::Pattern trait only implemented for &[T] ?

Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

KodrAus commented:
We discussed this a bit in the recent Libs meeting and felt that if we wanted to bring this forward and stabilize it ahead of reworking the pattern API to be non-`str` specific then we could reduce some churn by making the signatures generic on a trivial unstable trait that we replace with `Pattern` when we can.

Something like:

```rust
#[unstable] // We never stabilize this trait
pub trait SlicePattern {
    type Item;
    fn as_slice(&self) -> &[Self::Item];
}

impl<T> SlicePattern for [T] {
    type Item = T;

    fn as_slice(&self) -> &[Self::Item] {
        self
    }
}

impl<T> [T] {
    #[stable]
    pub fn strip_prefix<P: SlicePattern<Item = T>>(&self, prefix: P) -> Option<&[T]>;

    #[stable]
    pub fn strip_suffix<P: SlicePattern<Item = T>>(&self, suffix: P) -> Option<&[T]>;
}
```

We may still get breakage from inference failures, but bring the API closer to what we actually want in the long term, and only need to change the unstable backing trait.

What do you think?

ijackson commented:
Ashley Mannix writes ("Re: [rust-lang/rust] Stabilize slice::strip_prefix and slice::strip_suffix (#77853)"):
> We discussed this a bit in the recent Libs meeting and felt that if we wanted
> to bring this forward and stabilize it ahead of reworking the pattern API to be
> non-str specific then we could reduce some churn by making the signatures
> generic on a trivial unstable trait that we replace with Pattern when we can.

I thought this might be a possibility and was meaning to suggest it.

> We may still get breakage from inference failures, but bring the API closer to
> what we actually want in the long term, and only need to change the unstable
> backing trait.

If the inference break is a small enough problem that this approach
would be suitable for stabilisation, then great.  I will put this item
back on my list for "unblocked, next thing is to write code".

Ian.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.

ijackson commented:
Hi.  I have updated this now.  I wrote a bunch of stuff in the commit message which github likes to hide so here it is for your err delectation and delight?

```
Stablize slice::strip_prefix and strip_suffix, with SlicePattern

We hope later to extend `core::str::Pattern` to slices too, perhaps as
part of stabilising that.  We want to minimise the amount of type
inference breakage when we do that, so we don't want to stabilise
strip_prefix and strip_suffix taking a simple `&[T]`.

@KodrAus suggested the approach of introducing a new perma-unstable
trait, which reduces this future inference break risk.

I found it necessary to make two impls of this trait, as the unsize
coercion don't apply when hunting for trait implementations.

Since SlicePattern's only method returns a reference, and the whole
trait is just a wrapper for slices, I made the trait type be the
non-reference type [T] or [T;N] rather than the reference.  Otherwise
the trait would have a lifetime parameter.

I marked both the no-op conversion functions `#[inline]`.  I'm not
sure if that is necessary but it seemed at the very least harmless.
```

KodrAus reviewed (commented):
Thanks @ijackson! This looks like a good approach to me.

ijackson commented on library/core/src/slice/mod.rs:
```diff
@@ -3218,3 +3220,35 @@ impl<T> Default for &mut [T] {
     }
 }
+
+#[unstable(feature = "x", issue = "none")]
```
Thanks, LGTM.

Amanieu commented:
@rust-lang/libs could you fill in checkboxes in https://github.com/rust-lang/rust/pull/77853#issuecomment-709644329 considering the updated proposal with a trait bound?

dtolnay commented on library/core/src/slice/mod.rs:
```diff
@@ -1778,20 +1778,21 @@ impl<T> [T] {
-        let n = prefix.len();
+        // This function will need rewriting if and when SlicePattern becomes more sophisticated.
+        let prefix = prefix.as_slice();
+        let n = prefix.as_slice().len();
```
```suggestion
        let n = prefix.len();
```

ijackson replied:
Thanks, yes, obviously :-).

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/1612491350) [(plain)](https://github.com/rust-lang/rust/commit/36be696bfb9efceefa5e7dfb46f33d01b1b9b9f8/checks/1612491350/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.channel         := nightly
configure: rust.debug-assertions := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
    Checking hashbrown v0.9.0
    Checking object v0.22.0
    Checking miniz_oxide v0.4.0
    Checking addr2line v0.14.0
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
   --> library/std/src/sys/windows/path.rs:103:37
    |
103 |     match path.bytes().strip_prefix(prefix.as_bytes()) {
    |                                     ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `Sized` is not implemented for `[u8]`
error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `std`
error: could not compile `std`

To learn more, run the command again with --verbose.
command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "check" "--target" "i686-pc-windows-gnu" "-Zbinary-dep-depinfo" "-j" "16" "--release" "--locked" "--color" "always" "--features" "panic-unwind backtrace compiler-builtins-c" "--manifest-path" "/checkout/library/test/Cargo.toml" "--message-format" "json-render-diagnostics"
failed to run: /checkout/obj/build/bootstrap/debug/bootstrap check --target=i686-pc-windows-gnu --host=i686-pc-windows-gnu --all-targets
Build completed unsuccessfully in 0:00:30
```

</details>

ijackson commented:
Investigating.

ijackson commented:
How embarrassing.  I forgot the `?Sized` bound on the type parameters, so this new version didn't work when the pattern was a slice!  All the tests were for arrays, so the tests didn't spot it.  Luckily something on Windows was added in the meantime which used `strip_prefix` with an actual slice, so that found the problem

I have rebased onto master, and fixed this bug, and also added a doctest which passes a slice (playing about a bit with byte strings).  LMK if you would prefer some different test.

lzutao commented on library/core/src/slice/mod.rs:
```diff
@@ -3216,3 +3222,35 @@ impl<T> Default for &mut [T] {
+/// Patterns in slices - currently, only used by `strip_prefix` and `strip_suffix`.  At a future
+/// point, we hope to generalise `core::str::Pattern` (which at the time of writing is limited to
+/// `str`) to slices, and then this trait will be replaced or abolished.
+pub trait SlicePattern {
```
Should we seal this trait?

Amanieu replied:
I don't think it is necessary right now since the trait is still unstable.

Amanieu commented on library/core/src/slice/mod.rs:
```diff
@@ -3216,3 +3222,35 @@ impl<T> Default for &mut [T] {
+    fn as_slice(&self) -> &[Self::Item];
+}
+
+#[stable(feature = "slice_strip", since = "1.50.0")]
```
Stability attributes don't work on trait impls, I think you can just remove them.

jplatte replied:
They are being rendered by rustdoc, e.g. https://doc.rust-lang.org/core/num/struct.NonZeroU128.html#impl-TryFrom%3CNonZeroU128%3E-1

Amanieu commented:
@bors r+

bors commented:
:pushpin: Commit 8b2e79dbac324b703493113c9f823e63cf1df4e0 has been approved by `Amanieu`

<!-- @bors r=Amanieu 8b2e79dbac324b703493113c9f823e63cf1df4e0 -->
<!-- homu: {"type":"Approved","sha":"8b2e79dbac324b703493113c9f823e63cf1df4e0","approver":"Amanieu"} -->

bors commented:
:hourglass: Testing commit 8b2e79dbac324b703493113c9f823e63cf1df4e0 with merge 8f0b945cfcc3d084583bc27a7ed23b27b1246751...
<!-- homu: {"type":"BuildStarted","head_sha":"8b2e79dbac324b703493113c9f823e63cf1df4e0","merge_sha":"8f0b945cfcc3d084583bc27a7ed23b27b1246751"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1664257975)
Approved by: Amanieu
Pushing 8f0b945cfcc3d084583bc27a7ed23b27b1246751 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Amanieu","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/1664257975"},"merge_sha":"8f0b945cfcc3d084583bc27a7ed23b27b1246751"} -->

ijackson commented:
Yay :-).  Thanks all

lzutao commented on library/core/src/slice/mod.rs:
```diff
@@ -1778,19 +1778,24 @@ impl<T> [T] {
     #[must_use = "returns the subslice without modifying the original"]
-    #[unstable(feature = "slice_strip", issue = "73413")]
-    pub fn strip_prefix(&self, prefix: &[T]) -> Option<&[T]>
+    #[stable(feature = "slice_strip", since = "1.50.0")]
```
This should be 1.51 right ?

ijackson replied:
Err, perhaps.  I have just tried looking for git tags or something to tell me which Rust versions contain 8f0b945cfcc3d084583bc27a7ed23b27b1246751, the relevant commit, but there don't seem to be any tags for beta versions.  I guess that makes sense.  I found that `origin/beta` does not contain that commit, and the forge tells me beta is 1.50.  So should I send a MR to fix this ?

lzutao replied:
This PR is on 1.51.0 milestone according to rustbot.
Yes, someone could send another PR to update the stabilization version.

ijackson commented:
> This PR is on 1.51.0 milestone according to rustbot.
> Yes, someone could send another PR to update the stabilization version.

Willdo then, thanks.

-- 
Ian Jackson <ijackson@chiark.greenend.org.uk>   These opinions are my own.  

Pronouns: they/he.  If I emailed you from @fyvzl.net or @evade.org.uk,
that is a private address which bypasses my fierce spamfilter.
